File: EgoCapture/Services/Diagnostics/FOVEnumerationDiagnostic.swift
import Foundation
import AVFoundation
import CoreMedia
import UIKit

/// One-shot enumeration of every back-facing capture device + format, including
/// virtual (multi-camera) composites, for the purpose of answering the question
/// "what is the highest diagonal FOV this iPhone can deliver in video mode?".
///
/// Pure inspection: does not touch the recording pipeline, does not start
/// `AVCaptureSession`, and can be called at any time on the main thread.
enum FOVEnumerationDiagnostic {

    /// Runs the enumeration, writes a JSON artifact to `Documents/FOVDiagnostics/`,
    /// and returns the file URL. The URL is suitable for a `UIActivityViewController`
    /// share sheet or for out-of-band uploads.
    @discardableResult
    static func runAndSave() throws -> URL {
        let payload = build()
        let dir = try diagnosticsDirectory()
        let filename = "fov_enumeration_\(UIDevice.current.model.replacingOccurrences(of: " ", with: "_"))_\(timestampSuffix()).json"
        let url = dir.appendingPathComponent(filename)
        let data = try JSONSerialization.data(
            withJSONObject: payload,
            options: [.prettyPrinted, .sortedKeys]
        )
        try data.write(to: url, options: .atomic)
        logSummary(payload: payload, fileURL: url)
        return url
    }

    // MARK: - Payload builder

    /// Builds the full enumeration payload as a JSON-serializable dictionary.
    /// Runs synchronously — enumeration is cheap (a few ms).
    static func build() -> [String: Any] {
        let virtualTypes: [AVCaptureDevice.DeviceType] = [
            .builtInDualWideCamera,
            .builtInTripleCamera,
            .builtInDualCamera
        ]
        let physicalTypes: [AVCaptureDevice.DeviceType] = [
            .builtInUltraWideCamera,
            .builtInWideAngleCamera
        ]
        let discovery = AVCaptureDevice.DiscoverySession(
            deviceTypes: virtualTypes + physicalTypes,
            mediaType: .video,
            position: .back
        )

        var deviceEntries: [[String: Any]] = []
        var maxDFov: Double = 0
        var maxDFovRecord: [String: Any]? = nil

        for device in discovery.devices {
            var formatEntries: [[String: Any]] = []
            for format in device.formats {
                let dims = CMVideoFormatDescriptionGetDimensions(format.formatDescription)
                let maxFPS = format.videoSupportedFrameRateRanges.map { $0.maxFrameRate }.max() ?? 0
                guard maxFPS >= 30, dims.width > 0, dims.height > 0 else { continue }

                let hFov = Double(format.videoFieldOfView)
                let gdcFov = Double(format.geometricDistortionCorrectedVideoFieldOfView)
                let dFov = diagonalFov(horizontalDeg: hFov, width: Int(dims.width), height: Int(dims.height))
                let aspectNumeric = Double(dims.width) / Double(dims.height)
                let fpsRanges = format.videoSupportedFrameRateRanges.map {
                    ["min": $0.minFrameRate, "max": $0.maxFrameRate]
                }

                let entry: [String: Any] = [
                    "width": Int(dims.width),
                    "height": Int(dims.height),
                    "aspectRatio": String(format: "%.3f", aspectNumeric),
                    "horizontalFovDeg": hFov,
                    "geometricDistortionCorrectedFovDeg": gdcFov,
                    "gdcReducesFov": gdcFov > 0 && gdcFov < hFov - 0.01,
                    "diagonalFovDeg": dFov,
                    "fpsRanges": fpsRanges
                ]
                formatEntries.append(entry)

                if dFov > maxDFov {
                    maxDFov = dFov
                    maxDFovRecord = [
                        "deviceType": device.deviceType.rawValue,
                        "deviceLocalizedName": device.localizedName,
                        "width": Int(dims.width),
                        "height": Int(dims.height),
                        "horizontalFovDeg": hFov,
                        "diagonalFovDeg": dFov
                    ]
                }
            }

            formatEntries.sort { lhs, rhs in
                let ld = (lhs["diagonalFovDeg"] as? Double) ?? 0
                let rd = (rhs["diagonalFovDeg"] as? Double) ?? 0
                return ld > rd
            }

            var deviceInfo: [String: Any] = [
                "deviceType": device.deviceType.rawValue,
                "localizedName": device.localizedName,
                "isVirtualDevice": device.isVirtualDevice,
                "gdcSupported": device.isGeometricDistortionCorrectionSupported,
                "gdcEnabled": device.isGeometricDistortionCorrectionEnabled,
                "formats30fpsPlus": formatEntries,
                "formatCount": formatEntries.count
            ]
            if device.isVirtualDevice {
                deviceInfo["constituentDevices"] = device.constituentDevices.map {
                    [
                        "deviceType": $0.deviceType.rawValue,
                        "localizedName": $0.localizedName
                    ]
                }
            }
            deviceEntries.append(deviceInfo)
        }

        var payload: [String: Any] = [
            "schema": "kgeneye.fov_enumeration.v1",
            "capturedAtEpochMs": Date().timeIntervalSince1970 * 1000.0,
            "deviceModel": UIDevice.current.model,
            "systemName": UIDevice.current.systemName,
            "systemVersion": UIDevice.current.systemVersion,
            "idfvIdentifier": UIDevice.current.identifierForVendor?.uuidString ?? "unknown",
            "summary": [
                "backDeviceCount": deviceEntries.count,
                "maxDiagonalFovDeg": maxDFov,
                "specMinimumDiagonalFovDeg": 120.0,
                "meetsSpecMinimum": maxDFov >= 120.0,
                "maxDiagonalFovFormat": maxDFovRecord as Any
            ],
            "devices": deviceEntries
        ]

        // Strip NSNull that `as Any` may introduce when the record is missing.
        if maxDFovRecord == nil {
            var summary = payload["summary"] as! [String: Any]
            summary.removeValue(forKey: "maxDiagonalFovFormat")
            payload["summary"] = summary
        }

        return payload
    }

    // MARK: - Helpers

    private static func diagonalFov(horizontalDeg: Double, width: Int, height: Int) -> Double {
        guard horizontalDeg > 0, width > 0, height > 0 else { return 0 }
        let hRad = horizontalDeg * .pi / 180.0
        let aspect = Double(height) / Double(width)
        let diag = 2.0 * atan(sqrt(1.0 + aspect * aspect) * tan(hRad / 2.0))
        return diag * 180.0 / .pi
    }

    private static func diagnosticsDirectory() throws -> URL {
        let fm = FileManager.default
        let docs = try fm.url(
            for: .documentDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let dir = docs.appendingPathComponent("FOVDiagnostics", isDirectory: true)
        if !fm.fileExists(atPath: dir.path) {
            try fm.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private static func timestampSuffix() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd_HHmmss"
        f.timeZone = TimeZone(identifier: "UTC")
        return f.string(from: Date()) + "Z"
    }

    private static func logSummary(payload: [String: Any], fileURL: URL) {
        let summary = payload["summary"] as? [String: Any] ?? [:]
        let devices = payload["devices"] as? [[String: Any]] ?? []
        let maxD = (summary["maxDiagonalFovDeg"] as? Double) ?? 0
        let meets = (summary["meetsSpecMinimum"] as? Bool) ?? false
        print("[FOVDiagnostic] === FOV ENUMERATION ===")
        print("[FOVDiagnostic] Back devices: \(devices.count)")
        for d in devices {
            let type = d["deviceType"] as? String ?? "?"
            let virtual = (d["isVirtualDevice"] as? Bool) == true ? "virtual" : "physical"
            let count = d["formatCount"] as? Int ?? 0
            let top = (d["formats30fpsPlus"] as? [[String: Any]])?.first
            let topFov = (top?["diagonalFovDeg"] as? Double) ?? 0
            let topW = top?["width"] as? Int ?? 0
            let topH = top?["height"] as? Int ?? 0
            print("[FOVDiagnostic]   \(type) (\(virtual)) — \(count) fmts, top=\(topW)x\(topH) dFov=\(String(format: "%.1f°", topFov))")
        }
        print("[FOVDiagnostic] maxDiagonalFovDeg=\(String(format: "%.2f°", maxD)) | meetsSpec120=\(meets)")
        print("[FOVDiagnostic] Saved: \(fileURL.path)")
    }
}

File: android/app/src/main/java/com/kgeneye/eye/session/FovEnumerationDiagnostic.kt
package com.kgeneye.eye.session

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraMetadata
import android.media.MediaRecorder
import android.os.Build
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.atan
import kotlin.math.hypot
import kotlin.math.roundToInt

/**
 * One-shot enumeration of every back-facing physical camera + format, for
 * answering "what is the highest diagonal FOV this Android phone can deliver
 * in 30 fps+ video?" Mirrors iOS' `FOVEnumerationDiagnostic` so reports from
 * both platforms can be compared with the same backend tooling.
 *
 * Pure inspection: does not open a `CameraDevice`, does not touch the
 * recording pipeline, safe to call from the UI thread.
 *
 * Output JSON lands in `getExternalFilesDir("FOVDiagnostics")`, so users can
 * pick it up over USB (`adb pull`) or share via Files app.
 */
object FovEnumerationDiagnostic {

    private const val TAG = "FOVDiagnostic"
    private const val SCHEMA = "kgeneye.fov_enumeration.v1"
    private const val SPEC_MIN_DIAGONAL_FOV_DEG = 120.0

    /**
     * Runs the enumeration, writes a JSON artifact to
     * `<external-files>/FOVDiagnostics/`, and returns the file. Returns null
     * if the device has no `CameraManager` (e.g. some emulators).
     */
    fun runAndSave(context: Context): File? {
        val payload = build(context) ?: return null
        val dir = diagnosticsDirectory(context) ?: return null
        val deviceTag = "${Build.MANUFACTURER}_${Build.MODEL}"
            .replace(" ", "_")
            .replace("/", "_")
        val filename = "fov_enumeration_${deviceTag}_${timestampSuffix()}.json"
        val file = File(dir, filename)
        file.writeText(payload.toString(2))
        logSummary(payload, file)
        return file
    }

    /**
     * Builds the full enumeration payload as a JSON object. Runs synchronously
     * (a few ms even on cheap devices).
     */
    fun build(context: Context): JSONObject? {
        val manager = context.getSystemService(Context.CAMERA_SERVICE) as? CameraManager ?: return null
        val ids = try { manager.cameraIdList } catch (_: Throwable) { return null }

        val devices = JSONArray()
        var maxDFov = 0.0
        var maxDFovRecord: JSONObject? = null

        for (id in ids) {
            val chars = try { manager.getCameraCharacteristics(id) } catch (_: Throwable) { continue }
            val facing = chars.get(CameraCharacteristics.LENS_FACING)
            if (facing != CameraMetadata.LENS_FACING_BACK) continue

            val sensorSize = chars.get(CameraCharacteristics.SENSOR_INFO_PHYSICAL_SIZE)
            val focalLengths = chars.get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)
            val pixelArray = chars.get(CameraCharacteristics.SENSOR_INFO_PIXEL_ARRAY_SIZE)
            val configMap = chars.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
            val fpsRanges = chars.get(CameraCharacteristics.CONTROL_AE_AVAILABLE_TARGET_FPS_RANGES)

            val sensorWmm = sensorSize?.width?.toDouble() ?: 0.0
            val sensorHmm = sensorSize?.height?.toDouble() ?: 0.0
            val focalMm = focalLengths?.firstOrNull()?.toDouble() ?: 0.0
            val supports30 = fpsRanges?.any { it.upper >= 30 } == true
            val maxFps = fpsRanges?.maxOfOrNull { it.upper } ?: 0
            val minFps = fpsRanges?.minOfOrNull { it.lower } ?: 0

            val sizes = configMap?.getOutputSizes(MediaRecorder::class.java) ?: emptyArray()

            val formatEntries = JSONArray()
            for (size in sizes) {
                val w = size.width
                val h = size.height
                if (w <= 0 || h <= 0 || !supports30) continue

                val (hFov, dFov) = computeFov(sensorWmm, sensorHmm, focalMm, w, h)
                val aspectNumeric = w.toDouble() / h.toDouble()
                formatEntries.put(JSONObject().apply {
                    put("width", w)
                    put("height", h)
                    put("aspectRatio", String.format(Locale.US, "%.3f", aspectNumeric))
                    put("horizontalFovDeg", hFov)
                    // Android Camera2 has no GDC equivalent; surface the same key
                    // shape as iOS for backend symmetry.
                    put("geometricDistortionCorrectedFovDeg", hFov)
                    put("gdcReducesFov", false)
                    put("diagonalFovDeg", dFov)
                    put("fpsRanges", JSONArray().apply {
                        fpsRanges?.forEach { range ->
                            put(JSONObject().apply {
                                put("min", range.lower)
                                put("max", range.upper)
                            })
                        }
                    })
                })

                if (dFov > maxDFov) {
                    maxDFov = dFov
                    maxDFovRecord = JSONObject().apply {
                        put("cameraId", id)
                        put("deviceLocalizedName", localizedName(id, focalMm))
                        put("width", w)
                        put("height", h)
                        put("horizontalFovDeg", hFov)
                        put("diagonalFovDeg", dFov)
                    }
                }
            }

            // Sort descending by diagonal FOV.
            val sorted = JSONArray()
            (0 until formatEntries.length())
                .map { formatEntries.getJSONObject(it) }
                .sortedByDescending { it.optDouble("diagonalFovDeg", 0.0) }
                .forEach { sorted.put(it) }

            val deviceInfo = JSONObject().apply {
                put("cameraId", id)
                put("localizedName", localizedName(id, focalMm))
                // Camera2 has no virtual/composite device concept; always physical.
                put("isVirtualDevice", false)
                // Lens distortion correction support varies; report explicitly when known.
                put("lensDistortionCorrectionAvailable", chars.get(CameraCharacteristics.LENS_DISTORTION) != null)
                put("focalLengthMm", focalMm)
                put("sensorPhysicalSizeMm", JSONArray(listOf(sensorWmm, sensorHmm)))
                put(
                    "pixelArraySize",
                    if (pixelArray != null) JSONArray(listOf(pixelArray.width, pixelArray.height))
                    else JSONArray(),
                )
                put("minFPS", minFps)
                put("maxFPS", maxFps)
                put("formats30fpsPlus", sorted)
                put("formatCount", sorted.length())
            }
            devices.put(deviceInfo)
        }

        val summary = JSONObject().apply {
            put("backDeviceCount", devices.length())
            put("maxDiagonalFovDeg", maxDFov.round1())
            put("specMinimumDiagonalFovDeg", SPEC_MIN_DIAGONAL_FOV_DEG)
            put("meetsSpecMinimum", maxDFov >= SPEC_MIN_DIAGONAL_FOV_DEG)
            if (maxDFovRecord != null) put("maxDiagonalFovFormat", maxDFovRecord)
        }

        return JSONObject().apply {
            put("schema", SCHEMA)
            put("capturedAtEpochMs", System.currentTimeMillis())
            put("deviceManufacturer", Build.MANUFACTURER)
            put("deviceModel", Build.MODEL)
            put("deviceProduct", Build.PRODUCT)
            put("systemName", "Android")
            put("systemVersion", Build.VERSION.RELEASE)
            put("sdkInt", Build.VERSION.SDK_INT)
            put("summary", summary)
            put("devices", devices)
        }
    }

    private fun localizedName(cameraId: String, focalMm: Double): String {
        val role = when {
            focalMm in 0.01..2.5 -> "Ultra Wide"
            focalMm in 2.5..5.0 -> "Wide"
            focalMm > 5.0 -> "Telephoto"
            else -> "Rear"
        }
        return "$role (id=$cameraId)"
    }

    private fun computeFov(
        sensorWmm: Double, sensorHmm: Double, focalMm: Double,
        width: Int, height: Int,
    ): Pair<Double, Double> {
        if (focalMm <= 0.0 || sensorWmm <= 0.0) return 0.0 to 0.0
        val sensorAspect = sensorWmm / sensorHmm.coerceAtLeast(1e-6)
        val targetAspect = width.toDouble() / height.coerceAtLeast(1).toDouble()
        val activeWmm: Double
        val activeHmm: Double
        if (targetAspect >= sensorAspect) {
            activeWmm = sensorWmm
            activeHmm = sensorWmm / targetAspect
        } else {
            activeHmm = sensorHmm
            activeWmm = sensorHmm * targetAspect
        }
        val hFov = 2.0 * atan(activeWmm / (2.0 * focalMm))
        val diag = hypot(activeWmm, activeHmm)
        val dFov = 2.0 * atan(diag / (2.0 * focalMm))
        return Math.toDegrees(hFov).round1() to Math.toDegrees(dFov).round1()
    }

    private fun Double.round1(): Double = (this * 10.0).roundToInt() / 10.0

    private fun diagnosticsDirectory(context: Context): File? {
        val base = context.getExternalFilesDir(null) ?: context.filesDir
        val dir = File(base, "FOVDiagnostics")
        if (!dir.exists() && !dir.mkdirs()) {
            Log.w(TAG, "could not create ${dir.absolutePath}")
            return null
        }
        return dir
    }

    private fun timestampSuffix(): String {
        val fmt = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        return fmt.format(Date()) + "Z"
    }

    private fun logSummary(payload: JSONObject, file: File) {
        val summary = payload.optJSONObject("summary")
        val devices = payload.optJSONArray("devices") ?: JSONArray()
        val maxD = summary?.optDouble("maxDiagonalFovDeg", 0.0) ?: 0.0
        val meets = summary?.optBoolean("meetsSpecMinimum", false) ?: false
        Log.i(TAG, "=== FOV ENUMERATION ===")
        Log.i(TAG, "Back devices: ${devices.length()}")
        for (i in 0 until devices.length()) {
            val d = devices.getJSONObject(i)
            val name = d.optString("localizedName", "?")
            val count = d.optInt("formatCount", 0)
            val formats = d.optJSONArray("formats30fpsPlus")
            val top = if (formats != null && formats.length() > 0) formats.getJSONObject(0) else null
            val topFov = top?.optDouble("diagonalFovDeg", 0.0) ?: 0.0
            val topW = top?.optInt("width", 0) ?: 0
            val topH = top?.optInt("height", 0) ?: 0
            Log.i(TAG, "  $name — $count fmts, top=${topW}x$topH dFov=${"%.1f°".format(topFov)}")
        }
        Log.i(TAG, "maxDiagonalFovDeg=${"%.2f°".format(maxD)} | meetsSpec120=$meets")
        Log.i(TAG, "Saved: ${file.absolutePath}")
    }
}
